packages/flutter_tools/lib/src/asset.dart

```diff
@@ -912,8 +912,17 @@ class ManifestAssetBundle implements AssetBundle {
     Package? attributedPackage,
     List<String>? flavors,
   }) {
-    final String directoryPath = _fileSystem.path.join(
-        assetBase, assetUri.toFilePath(windows: _platform.isWindows));
+    final String directoryPath = (() {
+      try {
+        return _fileSystem.path
+          .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
+      } on UnsupportedError catch (e) {
+        throwToolExit(
+          'Unable to search for asset files in directory path "${assetUri.path}". '
+          'Please ensure that this is valid URI that points to a directory '
+          'that is available on the local file system.\nError details:\n$e');
+      }
+    })();
```

Comment:
I think you can safely avoid the immediately-invoked function expression because the compiler is smart and `throwToolExit()` has the return type `Never`, which guarantees it will never return.

```suggestion
    final String directoryPath;
    try {
      directoryPath = _fileSystem.path
        .join(assetBase, assetUri.toFilePath(windows: _platform.isWindows));
    } on UnsupportedError catch (e) {
      throwToolExit(
        'Unable to search for asset files in directory path "${assetUri.path}". '
        'Please ensure that this is valid URI that points to a directory '
        'that is available on the local file system.\nError details:\n$e');
    }
```

Reply:
I don't know what a URI is, I suspect most of our users don't either. Can we be more practical? What happened is that they either put an invalid path in their pubspec.yaml assets key or that directory exists but we for some other reason failed to parse the URI, right?